src/app/model/migrations/migrations/to0002/MigrationUtils0002.ts

```diff
@@ -0,0 +1,173 @@
+import { QueryClient } from '@tanstack/react-query';
+
+import { ActivityPipelineType } from '@app/abstract/lib';
+import { getAppletDetailsKey, getDataFromQuery } from '@app/shared/lib';
+
+import {
+  AppletDetailsDto,
+  AppletDetailsResponse,
+} from './MigrationDtoTypes0002';
+import {
+  RootStateFrom,
+  RootStateTo,
+  StoreProgressPayloadFrom,
+  StoreProgressPayloadTo,
+  StoreProgressTo,
+} from './MigrationReduxTypes0002';
+
+export type NotCompletedEntitiesTo = {
+  type: ActivityPipelineType;
+  appletId: string;
+  entityId: string;
+  eventId: string;
+  payload: StoreProgressPayloadTo;
+};
+
+export type NotCompletedEntitiesFrom = {
+  type: ActivityPipelineType;
+  appletId: string;
+  entityId: string;
+  eventId: string;
+  payload: StoreProgressPayloadFrom;
+};
+
+export const selectNotCompletedFlows = (
+  state: RootStateFrom,
+): NotCompletedEntitiesFrom[] => {
+  const inProgressApplets = state.applets.inProgress ?? {};
+  const result: NotCompletedEntitiesFrom[] = [];
+
+  const appletIds = Object.keys(inProgressApplets);
+
+  for (const appletId of appletIds) {
+    const progressEntities = inProgressApplets[appletId] ?? {};
+
+    const entityIds = Object.keys(progressEntities);
+
+    for (const entityId of entityIds) {
+      const progressEvents = progressEntities[entityId] ?? {};
+
+      const eventIds = Object.keys(progressEvents);
+
+      for (const eventId of eventIds) {
+        const payload = progressEvents[eventId] ?? {};
+
+        if (
+          !progressEvents[eventId] ||
+          payload.endAt ||
+          payload.type === ActivityPipelineType.Regular
+        ) {
+          continue;
+        }
+
+        result.push({
+          appletId,
+          entityId,
+          eventId,
+          type: payload.type,
+          payload,
+        });
+      }
+    }
+  }
+
+  return result;
+};
+
+export const selectNotCompletedActivities = (
+  state: RootStateFrom,
+): NotCompletedEntitiesFrom[] => {
+  const inProgressApplets = state.applets.inProgress ?? {};
+  const result: NotCompletedEntitiesFrom[] = [];
+
+  const appletIds = Object.keys(inProgressApplets);
+
+  for (const appletId of appletIds) {
+    const progressEntities = inProgressApplets[appletId] ?? {};
+    const entityIds = Object.keys(progressEntities);
+
+    for (const entityId of entityIds) {
+      const progressEvents = progressEntities[entityId] ?? {};
+
+      const eventIds = Object.keys(progressEvents);
+
+      for (const eventId of eventIds) {
+        const payload = progressEvents[eventId] ?? {};
+
+        if (
+          !progressEvents[eventId] ||
+          payload.endAt ||
+          payload.type !== ActivityPipelineType.Regular
+        ) {
+          continue;
+        }
+
+        result.push({
+          appletId,
+          entityId,
+          eventId,
+          type: payload.type,
+          payload,
+        });
+      }
+    }
+  }
+
+  return result;
+};
```

Comment:
I can see that these two functions `selectNotCompletedFlows` and `selectNotCompletedActivities` have little difference. Let's create a common function like this:
```
export const selectNotCompletedEntities = (
  state: RootStateFrom,
  typeFilter: 'flows' | 'activities'
): NotCompletedEntitiesFrom[] => {
  const inProgressApplets = state.applets.inProgress ?? {};
  const result: NotCompletedEntitiesFrom[] = [];

  const appletIds = Object.keys(inProgressApplets);

  for (const appletId of appletIds) {
    const progressEntities = inProgressApplets[appletId] ?? {};
    const entityIds = Object.keys(progressEntities);

    for (const entityId of entityIds) {
      const progressEvents = progressEntities[entityId] ?? {};

      const eventIds = Object.keys(progressEvents);

      for (const eventId of eventIds) {
        const payload = progressEvents[eventId] ?? {};

        if (
          !progressEvents[eventId] ||
          payload.endAt ||
          (typeFilter === 'flows' && payload.type === ActivityPipelineType.Regular) ||
          (typeFilter === 'activities' && payload.type !== ActivityPipelineType.Regular)
        ) {
          continue;
        }

        result.push({
          appletId,
          entityId,
          eventId,
          type: payload.type,
          payload,
        });
      }
    }
  }

  return result;
};

```